doc/iterations/2024.11/add_undo_redo_support.adoc

```diff
@@ -38,6 +38,34 @@ Pressing ctrl+y will redo the action that was previously undone.
 
 This should work regardless of the representation.
 
+=== Scenario (serialized representation)
+
+Serialized representation currently used in Sirius Web contains a layout part that is not contained in an EMF resource.
+
+On an unsynchronized representation the same scenarios as above should work.
+
+1. Several users are working on the same node 1
+- An user u1 resize a node from 10*10 to 100*100
+- An user u2 add sub nodes to the same node
+- If An user u1 press ctrl+z
+- *Expected*: the node will be not be resize to 10*10 if the sub nodes added by u2 do not fit inside, it will be resized to a default size.
+
+2. Several users are working on the same node 2
+- An user u1 move a node from 0,0 to 10,10
+- An user u2 move this node from 10,10 to 20,20
+- If An user u1 press ctrl+z
+- *Expected*: nothing will change since it would create a conflict with the change from u2
+- If An user u2 press ctrl+z
+- *Expected*: the node is moved back to 10,10 and u1 can press ctrl-z to move it back to 0,0
```

Comment:
So from this example, I conclude that you are advocating for a system in which:

- we would find the impact of a change done by a mutation in representation data
- we would store this change, for example, `move node X from 0,0 to 10,10`

and when we would try to undo it, we would:

- figure out the inverse of the change
- detect if it can be applied
- apply it

I agree with the approach, now let's consider some concrete use cases to challenge that:

### First use case

Let's consider that I have a portal, I execute a mutation to add a representation, after that the representation is added to the portal. We would need to keep somewhere that change. How would you compute the change done here? By analyzing the mutation? By analyzing the difference in the two versions of the portal? In this case, the input of the mutation contains everything relevant, it's easy.


### Second use case

Now let's consider that I have a class diagram with the following description (this is directly inspired by Eclipse SysON):

```
nodeDescription Class {
  domainType: "papaya::Class"
  semanticCandidatesExpression: "aql:self.types"
  synchronization: UNSYNCHRONIZED
  childNodeDescriptions: [AllAttributes]
}

nodeDescription AllAttributes {
  domainType: "papaya::Class"
  semanticCandidatesExpression: "aql:self"
  synchronization: UNSYNCHRONIZED
  childNodeDescriptions: [Attribute]
}

nodeDescription Attribute {
  domainType: "papaya::Attribute"
  semanticCandidatesExpression: "aql:self.attributes"
  synchronization: SYNCHRONIZED
}
```

Now that consider that I have a tool which will, thanks to a Java services and only a Java service:

- create a view for an existing class
- create an attribute in said class
- create a view for the attribute container AllAttributes

We have already seen how to record and revert the semantic changes, here the creation of an attribute. How would you detect the representation change that have been made? Here, there are three new nodes in the diagram:

- A Class unsynchronized node
- An AllAttribute unsynchronized node
- An Attribute node which is synchronized

Analyzing the input won't work since it will only tell you that the tool X has been called. On top of that, everything has been done in a Java service by the specifier so you don't know what happens.

You could navigate in the whole diagram data structure of the previous and new diagram and detect the differences (here you have three new nodes). From there, you would have to consider if those nodes are synchronized or not and if they are you would then keep in mind that some create view operation has been made. For this specific use case, you could also detect it thanks to the diagram event.

After this refresh, the diagram will be sent back to the frontend, which will compute the layout and send it back to the backend using a mutation. Let's ignore this part for now.

### Third use case

Now let's consider that a couple minutes later, I will move the newly created node. The frontend will send the full new layout to the backend and you will have to compute the difference. It can be done by analyzing the layout data of the previous and new diagram. The same issue would appear with portals but the mutation is straightforward so you could figure it from the input of the mutation alone.

How would you distinguish the layout operation from someone who has explicitly performed a layout change in the frontend from a layout update in response to the execution of a tool? I think it can be done but we would need to figure out how.